ext/reflection/php_reflection.c

```diff
@@ -3106,19 +3151,116 @@ ZEND_METHOD(ReflectionNamedType, isBuiltin)
 }
 /* }}} */
 
-static void append_type(zval *return_value, zend_type type) {
+/* {{{ Resolve a relative class type to a proper named type */
+ZEND_METHOD(ReflectionRelativeClassType, resolveToNamedType)
+{
+	reflection_object *intern;
+	type_reference *param;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		RETURN_THROWS();
+	}
+	GET_REFLECTION_OBJECT_PTR(param);
+
+	/* Is closure */
+	if (!Z_ISUNDEF(intern->obj)) {
+		/* Unbound closures can use relative class types */
+		if (!intern->ce) {
+			ZEND_ASSERT(!Z_ISUNDEF(intern->obj));
+			zend_throw_exception_ex(reflection_exception_ptr, 0,
+				"Cannot resolve relative class name for a closure");
+			RETURN_THROWS();
+		}
+
+		/* Attempt to resolve bound Closure */
+		zend_function *fptr = NULL;
+		zend_class_entry *called_scope = NULL;
+		zend_object *this_obj = NULL;
+
+		Z_OBJ_HANDLER(intern->obj, get_closure)(Z_OBJ(intern->obj), &called_scope, &fptr, &this_obj, /* check_only */ true);
```

Comment:
I think that you could avoid calling `->get_closure` twice here and in `reflection_type_factory`:

- Call `->get_closure` in `reflection_type_factory` and store `ce`. Set it to `called_scope` for MAY_BE_STATIC, and `fptr->common.scope` otherwise
- Then here, just use `intern->ce` in all cases.

An additional benefit is that it may be possible to avoid holding the closure in `->obj`. For the condition above, you could add a flag in `type_reference` to indicate this is in a closure context.